state-chain/pallets/cf-broadcast/src/migrations/v3.rs

```diff
@@ -0,0 +1,283 @@
+}
+pub struct Migration<T: Config<I>, I: 'static>(PhantomData<(T, I)>);
+
+impl<T: Config<I>, I: 'static> OnRuntimeUpgrade for Migration<T, I> {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		// Awaiting broadcast: take the broadcast attempt with the highest attempt.
+		let mut latest_attempt = BTreeMap::new();
+		old::AwaitingBroadcast::<T, I>::iter_keys().for_each(
+			|old::BroadcastAttemptId { broadcast_id, attempt_count }| {
+				if match latest_attempt.get(&broadcast_id) {
+					Some(attempt) => attempt_count > *attempt,
+					None => true,
+				} {
+					latest_attempt.insert(broadcast_id, attempt_count);
+				}
+			},
+		);
+
+		// Migrate data from old storage into new storage.
+		latest_attempt.into_iter().for_each(|(broadcast_id, attempt_count)| {
+			if let Some(attempt) = old::AwaitingBroadcast::<T, I>::get(old::BroadcastAttemptId {
+				broadcast_id,
+				attempt_count,
+			}) {
+				AwaitingBroadcast::<T, I>::insert(
+					broadcast_id,
+					TransactionSigningAttempt {
+						broadcast_attempt: BroadcastAttempt {
+							broadcast_id,
+							transaction_payload: attempt.broadcast_attempt.transaction_payload,
+							threshold_signature_payload: attempt
+								.broadcast_attempt
+								.threshold_signature_payload,
+							transaction_out_id: attempt.broadcast_attempt.transaction_out_id,
+						},
+						nominee: attempt.nominee,
+					},
+				);
+			}
+		});
+
+		// Failed broadcaster storage: Vec<ValidatorId> -> BTreeSet<ValidatorId>
+		FailedBroadcasters::<T, I>::translate(|_, failed_broadcasters: Vec<T::ValidatorId>| {
+			Some(BTreeSet::from_iter(failed_broadcasters))
+		});
+
+		let _ = BroadcastRetryQueue::<T, I>::translate(
+			|attempts: Option<Vec<old::BroadcastAttempt<T, I>>>| {
+				Some(
+					attempts
+						.unwrap_or_default()
+						.into_iter()
+						.map(|attempt: old::BroadcastAttempt<T, I>| BroadcastAttempt::<T, I> {
+							broadcast_id: attempt.broadcast_attempt_id.broadcast_id,
+							transaction_payload: attempt.transaction_payload,
+							threshold_signature_payload: attempt.threshold_signature_payload,
+							transaction_out_id: attempt.transaction_out_id,
+						})
+						.collect::<Vec<_>>(),
+				)
+			},
+		);
+
+		Weight::zero()
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		let broadcasts_with_failed_broadcasters = old::FailedBroadcasters::<T, I>::iter()
+			.map(|(broadcast_id, failed_broadcasters)| {
+				(broadcast_id, failed_broadcasters.into_iter().collect())
+			})
+			.collect::<Vec<_>>();
+		let broadcast_retry_queue = old::BroadcastRetryQueue::<T, I>::get()
+			.unwrap_or_default()
+			.into_iter()
+			.map(|attempt| attempt.broadcast_attempt_id.broadcast_id)
+			.collect::<Vec<_>>();
+		Ok(MigrationVerification::<T> {
+			broadcasts_with_failed_broadcasters,
+			broadcast_retry_queue,
+		}
+		.encode())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(state: Vec<u8>) -> Result<(), DispatchError> {
```

Comment:
The post_upgrade checks could ensure that the old storage values have been removed.

Reply:
we now check the old AwaitingBroadcast storage is cleared.
However for `FailedBroadcasters` and `BroadcastRetryQueue`, the old storage is linked with the new one (clearing old storage also clears the new storage) - I suspect this is due to the use of the "translate" call to migrate.